src/mono/mono/eventpipe/ep-rt-mono.c

```diff
@@ -2857,6 +2857,383 @@ ep_rt_mono_write_event_ee_startup_start (void)
+
+		char *wsz_name = target->s_name;
+		if (!wsz_name)
+		{
+			m_p_bulk_type_event_buffer[i_size++] = 0;
+			m_p_bulk_type_event_buffer[i_size++] = 0;
+		}
+		else
+		{
+			uint32_t name_size = (strlen(target->s_name) + 1) * sizeof(wchar_t);
+			memcpy(m_p_bulk_type_event_buffer + i_size, wsz_name, name_size);
+			i_size += strlen(wsz_name);
+		}
+
+		uint32_t *ptr_int = (uint32_t*)(m_p_bulk_type_event_buffer + i_size);
+		*ptr_int = target->c_type_parameters;
+		i_size += 4;
+
+		if (target->c_type_parameters > 0)
+		{
+			memcpy(m_p_bulk_type_event_buffer + i_size, target->rg_type_parameters, sizeof(uint64_t) * target->c_type_parameters);
+			i_size += sizeof(uint64_t) * target->c_type_parameters;
+		}
+	}
+
+	mono_trace (G_LOG_LEVEL_DEBUG, MONO_TRACE_DIAGNOSTICS, "Firing BulkTypeEvent!");
+	FireEtwBulkType(m_n_bulk_type_value_count,
+					n_clr_instance_id,
+					i_size,
+					m_p_bulk_type_event_buffer,
+					NULL,
+					NULL);
+
+	m_n_bulk_type_value_count = 0;
+	m_n_bulk_type_value_byte_count = 0;
+}
+
+//---------------------------------------------------------------------------------------
+//
+// ep_rt_mono_log_single_type batches a single type into the bulk type array and flushes
+// the array to ETW if it fills up. Most interaction with the type system (type analysis)
+// is done here. This does not recursively batch up any parameter types (arrays or generics),
+// but does add their unique identifiers to the rg_type_parameters array.
+// ep_rt_mono_log_type_and_parameters is responsible for initiating any recursive calls to
+// deal with type parameters.
+//
+// Arguments:
+//      type_id - Unique identifier for a mono type
+//
+// Return Value:
+//      Index into array of where this type got batched. -1 if there was a failure.
+
+uint32_t
+ep_rt_mono_log_single_type (intptr_t type_id)
+{
+    // If there's no room for another type, flush what we've got
+	if (m_n_bulk_type_value_count == k_max_count_type_values)
+		ep_rt_mono_fire_bulk_type_event();
+
+	EP_ASSERT (m_n_bulk_type_value_count < k_max_count_type_values);
+
+	BulkTypeValue *p_val = &m_rg_bulk_type_values[m_n_bulk_type_value_count];
+
+	MonoType *mono_type = (MonoType*)type_id;
+	MonoClass *klass = mono_class_from_mono_type_internal (mono_type);
+
+    // Clear out p_val before filling it out (array elements can get reused if there
+    // are enough types that we need to flush to multiple events).
+	memset(p_val->rg_type_parameters, 0, 100);
+	if (p_val->s_name)
+		p_val->s_name[0] = '\0';
+	p_val->c_type_parameters = 0;
+
+	// Initialize p_val fixed_sized_data
+	p_val->fixed_sized_data.type_id = (uint64_t)type_id;
+	p_val->fixed_sized_data.module_id = (uint64_t)m_class_get_image (klass);
+	p_val->fixed_sized_data.type_name_id = (0x00FFFFFF & m_class_get_type_token (klass)) | 0x02000000;  // dotnet-pgo ResolveMethodID expects type name ids with a 0x02 mask
+	p_val->fixed_sized_data.flags = 0;
+	p_val->fixed_sized_data.cor_element_type = (uint8_t)mono_type->type;
```

Comment:
This must match CoreCLR's logic, TypeHandle::GetInternalCorElementType(), it does some special handling of some types and enums as well.

Reply:
I initially saw that TypeDesc's GetInternalCorElementType returns a [`CorElementType`](https://github.com/dotnet/runtime/blob/694067c12626668074197d2f85f7babe6f01ac7a/src/coreclr/inc/corhdr.h#L870-L923) which is mostly in line with [`MonoTypeEnum`](https://github.com/dotnet/runtime/blob/694067c12626668074197d2f85f7babe6f01ac7a/src/native/public/mono/metadata/blob.h#L12-L51) except for their `ELEMENT_TYPE_MAX 0x22` and our `MONO_TYPE_ENUM 0x55`. Looks like `MethodTable`'s `GetInternalCorElementType` returns the same type but based off of a flag https://github.com/dotnet/runtime/blob/694067c12626668074197d2f85f7babe6f01ac7a/src/coreclr/vm/methodtable.cpp#L4865-L4917. `CorElementType` and `MonoTypeEnum` *seem* to be equivalent, but I dont know what would be mono's `ELEMENT_TYPE_MAX` or how CoreCLR handles enums like `MONO_TYPE_ENUM`.

Reply:
I guess this is mostly accurate, but looking the comments in TypeHandle::GetInternalCorElementType made me realize that we might need to see how that applies to Mono as well:

//    * Enumerated type have been normalized to the primitive type that underlies them (typically int)
//    * Value types that look like ints (which include RuntimeTypeHandles, etc), have been morphed to be
//        their underlying type (much like enumeration types.  See

Reply:
we have some `underlying_type` helpers in various places to handle enums.   In other places where we do a `switch` on `MonoType:type` there's usually a `handle_enum:` label and when we see `MONO_TYPE_VALUETYPE` we get the underlying type of the enum and do a `goto handle_enum`.

For example here:
https://github.com/dotnet/runtime/blob/729bf92e6e2f91aa337da9459bef079b14a0bf34/src/mono/mono/mini/aot-runtime-wasm.c#L22-L53

Reply:
Also `MONO_TYPE_ENUM 0x55` is used during metadata parsing (it's a marker in signatures I think... I don't remember and don't have my ECMA pdf handy), but we never have a `MonoType:type` of `MONO_TYPE_ENUM`

Reply:
Used `mono_type_get_underlying_type (mono_type)` to get any underlying type of enumerated types or value types.